PSPs/drafts/psp-32.md

```diff
@@ -0,0 +1,837 @@
+- Formula: Metadata fees are constructed around two constants (MetadataDepositBase and MetadataDepositPerByte). The formula for calculating the metadata store fees is: `MetadataDepositBase + MetadataDepositPerByte * (key.len() + value.len())`
+- Limited: The maximum length of data stored on-chain per token/asset is limited by `KeyLimit`, `ValueLimit` and `AttributeLimit` properties.
+- Validatable: You can define some contraints on-chain to validate the data stored on-chain.
+
+### Off-Chain Metadata
+
+For store any data that you don't need on-chain, such as images, names, description, etc.
+
+Characteristics:
+- Cheap: Can be stored in (URI, IPFS, etc).
+- on-chain reference: Must have a on-chain reference stored at `:metadata` token attribute key.
+- Not validatable: Not readable on-chain, consequently you can't perform any validation or logic using this data.
+
+There are currently no plans in place to restrict the metadata payload format.
+
+## Events
+
+```Rust
+
+pub enum Event<T: Config> {
+	/// A new asset was created by `owner`.
+	/// \[ asset_id, owner \]
+	AssetCreated(T::AssetId, T::AccountId),
+
+	/// The given asset was destroyed.
+	/// \[ asset_id]
+	AssetDestroyed(T::AssetId),
+
+	/// The `amount` balance of the `asset` was minted by owner.
+	/// \[owner, asset, amount\]
+	Minted(T::AccountId, T::AssetId, T::AssetBalance),
+
+	/// New tokens were minted on asset. 
+	/// \[owner, asset, from_token, to_token \]
+	TokensMinted(T::AccountId, T::AssetId, T::TokenId, T::TokenId),
+
+	/// `TransferredSingle` MUST emit when tokens are transferred.
+	/// The `from` argument MUST be the address of the holder whose balance is decreased.
+	/// The `to` argument MUST be the address of the recipient whose balance is increased.
+	/// The `asset_id` argument MUST be the token's asset type.
+	/// The `token_id` argument MUST be the token being transferred.
+	/// The `amount` argument MUST be the number of tokens by which the holder balance is decreased 
+	/// and the recipient balance is increased.
+	/// \[from, to, asset_id, token_id, amount\]
+	TransferredSingle(T::AccountId, T::AccountId, T::AssetId, T::TokenId, T::Balance),
+
+	/// `AssetAttributeSet` MUST emit when asset attribute is set.
+	/// \[asset_id, attribute key, attribute value\]
+	AssetAttributeSet(T::AssetId, T::AttributeKey, T::AttributeValue),
+
+	/// `AssetAttributeCleared` MUST emite when the asset attribute is cleared.
+	/// \[asset_id, key\]
+	AssetAttributeCleared(T::AssetId, T::AttributeKey),
+
+	/// `TokenAttributeSet` MUST emit when the token attribute is set.
+	/// \[asset_id, token_id, key, value\]
+	TokenAttributeSet(T::AssetId, T::TokenId, T::AttributeKey, T::AttributeValue),
+
+	/// `TokenAttributeCleared` MUST emit when the token attribute is cleared.
+	/// \[asset_id, token_id, key\]
+	TokenAttributeCleared(T::AssetId, T::TokenId, T::AttributeKey),
+}
+```
+
+## Extension: Batch
```

Comment:
maybe this should be in another meta operation standard?

Reply:
What is a meta operation standard? @xlc Can you elaborate?

Reply:
I mean this doesn't really needs to be token related. For features like batch transaction, proxy, multisig etc, they could be in another standard for those utility/meta operations.

Reply:
You might be right that that is preferable.
Note, though, that the proposed batch transactions here allow for optimizations (e.g. optimized batch minting of NFTs) which generic batches (e.g. via `pallet_utilitly::batch`) would not.